File: src/native/androidProject/app/src/main/java/io/yourname/androidproject/camera/TorchController.kt
package io.yourname.androidproject.camera

import android.content.Context
import android.util.Log
import androidx.camera.core.Camera
import androidx.core.content.ContextCompat

/**
 * Manages torch (flashlight) state.
 * Guards against front-camera torch calls.
 * Fires ON_TORCH_CHANGED via [onTorchChanged] callback only after the
 * hardware operation succeeds (ListenableFuture resolved on main thread).
 */
class TorchController(
    private val context: Context,
    private val onTorchChanged: (enabled: Boolean) -> Unit
) {

    private val TAG = "TorchController"

    private var camera: Camera? = null
    private var currentFacing: String = "back"

    fun attachCamera(cam: Camera, facing: String) {
        camera = cam
        currentFacing = facing
    }

    fun detachCamera() {
        camera = null
    }

    fun setTorch(on: Boolean) {
        val cam = camera ?: run {
            Log.w(TAG, "setTorch($on) — camera is null, skipping")
            return
        }
        if (currentFacing == "front") {
            Log.w(TAG, "setTorch($on) — front camera has no torch, ignoring")
            return
        }
        Log.d(TAG, "setTorch($on)")
        cam.cameraControl.enableTorch(on).addListener(
            {
                Log.d(TAG, "enableTorch($on) resolved — notifying")
                onTorchChanged(on)
            },
            ContextCompat.getMainExecutor(context)
        )
    }

    /** Called after every bindCamera — torch always resets to off on session start. */
    fun notifyReset() {
        onTorchChanged(false)
        Log.d(TAG, "Torch reset to off (new session)")
    }
}

File: src/native/iosnativeWebView/Sources/Core/Camera/TorchController.swift
import Foundation
import AVFoundation
import os

private let logger = Logger(subsystem: Bundle.main.bundleIdentifier ?? "com.app", category: "TorchController")

/// Manages torch (flashlight) state.
/// Guards against front-camera torch calls.
/// Fires ON_TORCH_CHANGED via onTorchChanged callback.
class TorchController {

    private var device: AVCaptureDevice?
    private var currentFacing: String = "back"
    private let onTorchChanged: (Bool) -> Void

    init(onTorchChanged: @escaping (Bool) -> Void) {
        self.onTorchChanged = onTorchChanged
    }

    func attachDevice(_ device: AVCaptureDevice, facing: String) {
        self.device = device
        self.currentFacing = facing
    }

    func detachDevice() {
        device = nil
    }

    func setTorch(_ on: Bool) {
        guard let device else {
            logger.warning("setTorch(\(on)) — device is nil, skipping")
            return
        }
        guard currentFacing == "back" else {
            logger.warning("setTorch(\(on)) — front camera has no torch, ignoring")
            return
        }
        guard device.hasTorch else {
            logger.warning("setTorch(\(on)) — device has no torch")
            return
        }

        do {
            try device.lockForConfiguration()
            device.torchMode = on ? .on : .off
            device.unlockForConfiguration()
            onTorchChanged(on)
            logger.debug("setTorch(\(on))")
        } catch {
            logger.error("setTorch(\(on)) failed: \(error.localizedDescription)")
        }
    }

    /// Called after every session bind — torch always resets to off on session start.
    func notifyReset() {
        onTorchChanged(false)
        logger.debug("Torch reset to off (new session)")
    }
}
